// @toptal/picasso/Badge/Badge.js
import React, { Children, forwardRef } from 'react';
import { makeStyles } from '@material-ui/core/styles';
import cx from 'classnames';
import { Badge as MuiBadge } from '@material-ui/core';
import styles from './styles';
const useStyles = makeStyles(styles, { name: 'PicassoBadge' });
const thresholds = {
    small: 9,
    medium: 99,
    large: 99,
};
export const Badge = forwardRef(function Badge({ children, style, variant = 'white', size = 'large', content, className, max, 'data-testid': testId, }, ref) {
    const classes = useStyles();
    const hasChildren = Children.count(children) > 0;
    return (React.createElement(MuiBadge, { ref: ref, style: style, "data-testid": testId, badgeContent: content, max: max || thresholds[size], className: className, showZero: true, classes: {
            badge: cx(classes.root, classes[variant], classes[size], {
                [classes.static]: !hasChildren,
            }),
        }, overlap: 'rectangular' }, children));
});
Badge.defaultProps = {
    variant: 'white',
    size: 'large',
};
Badge.displayName = 'Badge';
export default Badge;
