// @toptal/picasso/Badge/styles.js
import { rem } from '@toptal/picasso-shared';
import { createStyles } from '@material-ui/core/styles';
const getSizeProps = (unitInRem) => ({
    borderRadius: unitInRem,
    height: unitInRem,
    minWidth: unitInRem,
});
export default ({ palette, typography }) => createStyles({
    root: {
        borderWidth: '1px',
        borderStyle: 'solid',
        fontSize: rem('10px'),
        lineHeight: rem('12px'),
        padding: '0 1px',
        fontWeight: typography.fontWeights.semibold,
    },
    static: {
        position: 'unset',
        transform: 'unset',
    },
    white: {
        background: palette.common.white,
        color: palette.grey.dark,
        borderColor: palette.grey.light2,
    },
    red: {
        color: palette.common.white,
        borderColor: palette.red.main,
        backgroundColor: palette.red.main,
    },
    small: Object.assign({ lineHeight: rem('10px') }, getSizeProps(rem('12px'))),
    medium: Object.assign({}, getSizeProps(rem('16px'))),
    large: Object.assign({ padding: '0 3px' }, getSizeProps(rem('20px'))),
});
