// @toptal/picasso/UserBadge/UserBadge.js
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
/* eslint-disable complexity */
import React, { forwardRef } from 'react';
import { makeStyles } from '@material-ui/core/styles';
import cx from 'classnames';
import Avatar from '../Avatar';
import Typography from '../Typography';
import Container from '../Container';
import styles from './styles';
const useStyles = makeStyles(styles, {
    name: 'PicassoUserBadge',
});
export const UserBadge = forwardRef(function UserBadge(props, ref) {
    const { avatar, name, renderName, size, title, renderTitle, invert, center, children, className, style, 
    // Avoid passing external classes inside the rest props
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    classes: externalClasses } = props, rest = __rest(props, ["avatar", "name", "renderName", "size", "title", "renderTitle", "invert", "center", "children", "className", "style", "classes"]);
    const classes = useStyles(props);
    const UserBadgeAvatar = React.isValidElement(avatar) ? (avatar) : (React.createElement(Avatar, { className: classes.avatar, name: name, size: size, src: avatar }));
    // if 'auto' then center if children are null
    const shouldCenter = center === true || (center === 'auto' && !children);
    const alignItems = shouldCenter ? 'center' : 'flex-start';
    let userTitle;
    if (title) {
        userTitle = renderTitle ? (renderTitle(title, invert)) : (React.createElement(Typography, { inline: true, invert: invert, className: classes.title, size: 'medium' }, title));
    }
    const userName = renderName ? (renderName(name, invert)) : (React.createElement(Typography, { className: classes.name, inline: true, variant: 'heading', size: 'small', invert: invert }, name));
    return (React.createElement(Container, Object.assign({}, rest, { ref: ref, flex: true, alignItems: alignItems, className: cx(classes.root, className), style: style }),
        UserBadgeAvatar,
        React.createElement(Container, { flex: true, direction: 'column', className: classes.infoContainer },
            React.createElement(Container, { flex: true, alignItems: 'center' },
                userName,
                userTitle),
            children)));
});
UserBadge.defaultProps = {
    center: 'auto',
    invert: false,
    size: 'xsmall',
};
UserBadge.displayName = 'UserBadge';
export default UserBadge;
