// @toptal/picasso/Grid/Grid.js
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
import React, { forwardRef } from 'react';
import { makeStyles } from '@material-ui/core/styles';
import { Grid as MUIGrid, } from '@material-ui/core';
import styles from './styles';
const humanToMUISpacing = (spacing) => {
    /** Material Design margins and columns follow an 8px square baseline grid */
    return (spacing / 8);
};
const useStyles = makeStyles(styles, {
    name: 'PicassoGrid',
});
// eslint-disable-next-line react/display-name
export const Grid = forwardRef(function Grid(props, ref) {
    const { children, spacing = 32, direction, alignItems, justifyContent, wrap, className, style } = props, rest = __rest(props, ["children", "spacing", "direction", "alignItems", "justifyContent", "wrap", "className", "style"]);
    const classes = useStyles();
    return (React.createElement(MUIGrid, Object.assign({}, rest, { ref: ref, container: true, spacing: humanToMUISpacing(spacing), direction: direction, alignItems: alignItems, justifyContent: justifyContent, wrap: wrap, classes: classes, className: className, style: style }), children));
});
Grid.defaultProps = {
    alignItems: 'flex-start',
    direction: 'row',
    justifyContent: 'flex-start',
    spacing: 32,
    wrap: 'wrap',
};
Grid.displayName = 'Grid';
export default Grid;
