// @toptal/picasso/AlertInline/AlertInline.js
import React, { forwardRef } from 'react';
import { makeStyles } from '@material-ui/core';
import Container from '../Container';
import Typography from '../Typography';
import { ExclamationSolid16, DoneSolid16, InfoSolid16 } from '../Icon';
import styles from './styles';
const icons = {
    red: React.createElement(ExclamationSolid16, { color: 'red' }),
    green: React.createElement(DoneSolid16, { color: 'dark-green' }),
    blue: React.createElement(InfoSolid16, { color: 'light-blue' }),
    yellow: React.createElement(ExclamationSolid16, { color: 'yellow' }),
};
const useStyles = makeStyles(styles, {
    name: 'PicassoAlertInline',
});
export const AlertInline = forwardRef(function Alert(props, ref) {
    const classes = useStyles();
    const { variant, children, className } = props;
    const icon = icons[variant];
    let typographyColor = variant;
    // to comply with the designs
    if (variant === 'blue') {
        typographyColor = 'light-blue';
    }
    return (React.createElement(Container, { inline: true, flex: true, ref: ref, className: classes.root },
        React.createElement(Container, { right: 'xsmall', flex: true, alignItems: 'center', className: classes.iconWrapper }, icon),
        React.createElement(Typography, { size: 'small', as: 'div', weight: 'semibold', color: typographyColor, className: className }, children)));
});
AlertInline.defaultProps = {
    variant: 'yellow',
};
AlertInline.displayName = 'AlertInline';
export default AlertInline;
